src/Processors/QueryPlan/Optimizations/optimizeJoinByLayers.cpp

```diff
@@ -0,0 +1,412 @@
+        join_step->keepLeftPipelineInOrder();
+}
+
+void optimizeJoinByLayers(QueryPlan::Node & root)
+{
+    struct Result
+    {
+        JoinsAndSourcesWithCommonPrimaryKeyPrefix joins;
+        ActionsDAG dag; /// For the leftmost source
+    };
+
+    struct Frame
+    {
+        const QueryPlan::Node * node;
+        size_t next_child_to_process = 0;
+        std::vector<std::optional<Result>> results{};
+    };
+
+    std::optional<Result> result;
+    std::stack<Frame> stack;
+    stack.push({&root});
+
+    while (!stack.empty())
+    {
+        auto & frame = stack.top();
+        if (frame.next_child_to_process > 0)
+            frame.results.push_back(std::move(result));
+
+        result = {};
+
+        if (frame.next_child_to_process < frame.node->children.size())
+        {
+            stack.push({frame.node->children[frame.next_child_to_process]});
+            ++frame.next_child_to_process;
+            continue;
+        }
+
+        if (auto * join_step = typeid_cast<JoinStep *>(frame.node->step.get()))
+        {
+            const auto & join = join_step->getJoin();
+
+            // std::cerr << "Processing Join\n";
+            // WriteBufferFromOwnString out;
+            // IQueryPlanStep::FormatSettings settings{out};
+            // join_step->describeActions(settings);
+            // std::cerr << out.stringView() << std::endl;
+
+            auto * hash_join = typeid_cast<HashJoin *>(join.get());
+            auto * concurrent_hash_join = typeid_cast<ConcurrentHashJoin *>(join.get());
+            auto * full_sorting_merge_join = typeid_cast<FullSortingMergeJoin *>(join.get());
+            bool is_algo_supported = hash_join || concurrent_hash_join || full_sorting_merge_join;
+
+            bool can_split_left_table = frame.results.front() != std::nullopt && is_algo_supported && !join->hasDelayedBlocks();
+            // std::cerr << "can_split_left_table " << can_split_left_table << std::endl;
+
+            const auto & table_join = join->getTableJoin();
+            auto kind = table_join.kind();
+            auto strictness = table_join.strictness();
+            const auto & clauses = table_join.getClauses();
+
+            bool can_split_join = frame.results.back() != std::nullopt && can_split_left_table
+                && (isLeft(kind) || isRight(kind) || isInner(kind) || isFull(kind))
+                && strictness != JoinStrictness::Asof
+                && clauses.size() == 1;
+
+            // std::cerr << "can_split_join " << can_split_join << std::endl;
+
+            JoinStep::PrimaryKeySharding sharding;
+            if (can_split_join)
+            {
+                // std::cerr << frame.results.front()->dag.dumpDAG() << std::endl;
+                // std::cerr << frame.results.back()->dag.dumpDAG() << std::endl;
+
+                sharding = findCommonPromaryKeyPrefixByJoinKey(
+                    frame.results.front()->joins.sources.front(), frame.results.front()->dag,
```

Comment:
Why do we take particulary front source? Is it left-most source?

Reply:
yes, apparently, it's the left-most. It's because that the right one we fill in the memory and can't enable partitioning (at least for now).
Also, technically, we can allow reordering in the future somehow based on sharding. But now it seems to be too complicated.